src/tracker.c

```diff
@@ -299,8 +299,9 @@ static void goto_order(int pos) {
 	initialize_state();
 	for (i = 0; i < pos; i++) {
 		int ch;
-		for (ch = 0; ch < 8 && state.chan[ch].ptr != NULL; ch++)
-		// an empty pattern will loop forever, so skip it
+		// Determine number of null tracks
+		for (ch = 0; ch < 8 && state.chan[ch].ptr == NULL; ch++) { /* do nothing */ }
+		// If some tracks were not null, then simulate them without generating audio.
 		if (ch != 8) {
 			while (do_cycle_no_sound(&state));
 		}
```

Comment:
It looks like this change will prevent `do_cycle_no_sound` to always be skipped for all channels since `ch` will always be 8 by then which may cause other bugs for tracks other than the attached bug ebm. It looks like the condition to skip null tracks was mistakenly put into the for condition when it should be within the loop itself. It probably doesn't help that the loop is not indented which obscures what the loop is doing. I suggest we change it instead to this which still solves the problem and retains the functionality for non-null channels and also makes it clearer what's in the loop.
```suggestion
		for (ch = 0; ch < 8; ch++) {
			// simulate playing the channel without actually generating audio, but skip null channels
			if (state.chan[ch].ptr != NULL) {
				while (do_cycle_no_sound(&state));
			}
		}
```

Reply:
> It looks like this change will prevent `do_cycle_no_sound` to always be skipped for all channels since `ch` will always be 8

I'm not sure exactly what this is trying to say. Could you give an example such as "channel 0 and 2-7 are null, channel 1 is non-null"?

> It looks like the condition to skip null tracks was mistakenly put into the for condition when it should be within the loop itself. It probably doesn't help that the loop is not indented which obscures what the loop is doing.

To be honest, I can't really speak to what the original code was trying to do - I pretty much changed the logic to what I thought it should be.

> I suggest we change it instead to this which still solves the problem and retains the functionality for non-null channels and also makes it clearer what's in the loop.

I think this has another issue that is also in the original code. It's my understanding that calling `do_cycle_no_sound` in a loop like that basically steps through the whole pattern for all channels, not just whichever one we are looking at right not - note that `ch` is not passed into the function call. I don't think this would cause any actual issues tbh because calling it again after the current pattern has been run all the way through probably just returns false without doing anything, but I think it would be better to stop after we've run it all the way through once.

I think that the easiest way to do that would be to add a `break` after line 305 in your suggested change. Then when we encountered a non-null track in the current pattern, it will simulate the whole pattern and then stop checking other tracks because we already took care of the whole thing.

Reply:
Oh, I see now. From your explanation it makes more sense now that a semicolon was missing at the end of the for loop and it was unintentionally looping over the `if` rather than unintentionally keeping it unindented. But yeah, I think using a loop body and a break makes it much more clear what's going on.